modules/47_regipol/regiCarbonPrice/postsolve.gms

```diff
@@ -256,18 +256,29 @@ loop((ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47)$pm_emiMktTarget(
         if(iteration.val eq p47_slopeReferenceIteration_iter(iteration,ttot,ext_regi),
-          pm_factorRescaleemiMktCO2Tax(ttot,ttot2,ext_regi,emiMktExt) = (1+pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt)) ** 2;
+          pm_factorRescaleemiMktCO2Tax(ttot,ttot2,ext_regi,emiMktExt) = power(1+pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt), 2);
 ***     else if for the extreme case of a perfect match with no change between the two iterations emisssion taxes used in the slope calculation, in order to avoid a division by zero error assume the rescale factor based on remaining deviation
         elseif((pm_taxemiMkt_iteration(iteration,ttot2,regi,emiMkt) eq pm_taxemiMkt_iteration(iteration2,ttot2,regi,emiMkt))),
-          pm_factorRescaleemiMktCO2Tax(ttot,ttot2,ext_regi,emiMktExt) = (1+pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt)) ** 2;
+          pm_factorRescaleemiMktCO2Tax(ttot,ttot2,ext_regi,emiMktExt) = power(1+pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt), 2);
+***     else if emission tax variation in relation to base iteration is very small, assume the rescale factor based on the remaining deviation to avoid very slow improvements 
+        elseif(abs((pm_taxemiMkt_iteration(iteration,ttot2,regi,emiMkt) - pm_taxemiMkt_iteration(iteration2,ttot2,regi,emiMkt))) lt 1e-2),
+          pm_factorRescaleemiMktCO2Tax(ttot,ttot2,ext_regi,emiMktExt) = power(1+pm_emiMktTarget_dev(ttot,ttot2,ext_regi,emiMktExt), 2);
 ***     else, use previous iteration information to define rescale factor  
         else
-***       calculate the slope of previous iterations mitigation levels when compared to relative price difference          
-          p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt) =
+***       calculate the slope of previous iterations mitigation levels when compared to relative price difference    
+          p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt) = 
             (p47_emiMktCurrent_iter(iteration,ttot,ttot2,ext_regi,emiMktExt) - p47_emiMktCurrent_iter(iteration2,ttot,ttot2,ext_regi,emiMktExt))
             /
-            (pm_taxemiMkt_iteration(iteration,ttot2,regi,emiMkt) - pm_taxemiMkt_iteration(iteration2,ttot2,regi,emiMkt))
-          ;
+            (pm_taxemiMkt_iteration(iteration,ttot2,regi,emiMkt) - pm_taxemiMkt_iteration(iteration2,ttot2,regi,emiMkt));
+***       if the slope is positive, recalculate the slope based on the initial iteration instead of the reference one because we assume a trade-off between tax and emission levels 
+          if(p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt) gt 0, 
+            p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt) = 
+              (p47_emiMktCurrent_iter(iteration,ttot,ttot2,ext_regi,emiMktExt) - p47_emiMktCurrent_iter("1",ttot,ttot2,ext_regi,emiMktExt))
+              /
+              (pm_taxemiMkt_iteration(iteration,ttot2,regi,emiMkt) - pm_taxemiMkt_iteration("1",ttot2,regi,emiMkt));
```

Comment:
This is a first attempt to get a more reasonable slope, if the normal calculation does not provide a reliable value. 

- Usually, iteration 1 has lower carbon prices than later iterations because they are initialized at lower levels, which makes the slope denominator positive, and iteration 1 has more emissions than the desirable targets, which causes the slope numerator to be negative. As a result the slope should reflect the trade-off between tax and emissions, which we are trying to enforce here.
- Also, calculating the slope using iteration 1 usually gives a conservative value for the slope because we can mitigate emissions at much lower carbon prices firstly.

But more importantly, if the slope using iteration 1 is not reasonable enough, I overwrite it with the next line of the code which actually clamps the slope at -0.3, instead of your suggested 0.1 value.

```
***       clamp negative slopes to avoid extreme changes on a single iteration (avoid corner cases where other parts of the model changes causing undesirable fluctuations on the calculated slope)
          p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt) = min(-0.3, p47_factorRescaleSlope(ttot,ttot2,ext_regi,emiMktExt
```